firebase-firestore/src/main/java/com/google/firebase/firestore/FirebaseFirestore.java

```diff
@@ -174,7 +174,16 @@ public void setFirestoreSettings(@NonNull FirebaseFirestoreSettings settings) {
   }
 
   private void ensureClientConfigured() {
-    if (client == null) {
+    // We do not bother to synchronized on client as we do not touch it if it is already set.
+    if (client != null) {
+      return;
+    }
+    synchronized (settings) {
+      // Check this again under synchronization.
```

Comment:
Double-checked locking does not work unless `client` is volatile.

The `settings` object isn't a good choice for synchronization because it's non-final (and is reassigned) so it's conceivable paths could attempt to lock different settings objects.

Better to synchronize on something like databaseId which is final and a private object.